src/cmdline/mod.rs

```diff
@@ -90,117 +95,294 @@ fn valid_element(s: &str) -> Result<()> {
+        <String as UpdateArgs>::is_empty(&self.0)
+    }
+
+    fn insert_string<T: AsRef<str>>(&mut self, slug: T) -> Result<()> {
+        self.0.insert_string(slug)
+    }
+
+    fn as_string(&self) -> String {
+        <String as UpdateArgs>::as_string(&self.0)
+    }
+}
+
+/// A builder for a kernel command line string that validates the string
+/// as it's being built.
 ///
 /// # Examples
 ///
 /// ```rust
 /// # use linux_loader::cmdline::*;
 /// # use std::ffi::CString;
-/// let cl = Cmdline::new(100);
-/// let cl_cstring = CString::new(cl).unwrap();
-/// assert_eq!(cl_cstring.to_str().unwrap(), "");
+/// let mut cl = Cmdline::new(10).unwrap();
+/// cl.insert_string("foobar", CmdlineArgType::BootArg).unwrap();
+/// assert_eq!(cl.as_cstring().unwrap().as_bytes_with_nul(), b"foobar\0");
 /// ```
-#[derive(Clone, Debug)]
+#[derive(Clone, Eq, Ord, PartialOrd, Debug, Default)]
 pub struct Cmdline {
-    line: String,
+    boot_args: BootArgs,
+    init_args: InitArgs,
     capacity: usize,
 }
 
 impl Cmdline {
-    /// Constructs an empty [`Cmdline`] with the given capacity, including the nul terminator.
-    ///
-    /// # Arguments
-    ///
-    /// * `capacity` - Command line capacity. Must be greater than 0.
+    /// Constructs an empty [`Cmdline`].
     ///
     /// # Examples
     ///
     /// ```rust
     /// # use linux_loader::cmdline::*;
-    /// let cl = Cmdline::new(100);
+    /// let cl = Cmdline::new(10).unwrap();
     /// ```
     /// [`Cmdline`]: struct.Cmdline.html
-    pub fn new(capacity: usize) -> Cmdline {
-        assert_ne!(capacity, 0);
-        Cmdline {
-            line: String::with_capacity(capacity),
-            capacity,
+    pub fn new(capacity: usize) -> Result<Cmdline> {
+        if capacity == 0 {
+            return Err(Error::InvalidCapacity);
         }
+
+        Ok(Cmdline {
+            boot_args: BootArgs(String::new()),
+            init_args: InitArgs(String::new()),
+            capacity,
+        })
     }
 
-    fn has_capacity(&self, more: usize) -> Result<()> {
-        let needs_space = if self.line.is_empty() { 0 } else { 1 };
-        if self.line.len() + more + needs_space < self.capacity {
+    /// Tries to set a new capacity of the `Cmdline`. It fails when `new_capacity`
+    /// is smaller than the C complatible representation of the current `Cmdline`
+    /// (which requires to be null terminated)
+    ///
+    /// # Arguments
+    ///
+    /// * `new_capacity` - New capacity to be set..
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use linux_loader::cmdline::*;
+    /// let mut cl = Cmdline::new(3).unwrap();
+    /// assert!(cl.insert_string("foo", CmdlineArgType::BootArg).is_err());
+    /// cl.set_capacity(4).unwrap();
+    /// assert!(cl.insert_string("foo", CmdlineArgType::BootArg).is_ok());
+    /// assert_eq!(cl.set_capacity(3), Err(Error::InvalidCapacity));
+    /// ```
+    pub fn set_capacity(&mut self, new_capacity: usize) -> Result<()> {
```

Comment:
What use case does this function serve? We can already set the capacity in the constructor. When would we need to update that?

Reply:
removing this as discussed